In [186]:
import numpy as np
import pandas as pd
import sklearn.datasets
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import recall_score, precision_score
pd.options.display.float_format ='{:,.2f}'.format

In [187]:
data = sklearn.datasets.load_wine()
data

{'data': array([[1.423e+01, 1.710e+00, 2.430e+00, ..., 1.040e+00, 3.920e+00,
         1.065e+03],
        [1.320e+01, 1.780e+00, 2.140e+00, ..., 1.050e+00, 3.400e+00,
         1.050e+03],
        [1.316e+01, 2.360e+00, 2.670e+00, ..., 1.030e+00, 3.170e+00,
         1.185e+03],
        ...,
        [1.327e+01, 4.280e+00, 2.260e+00, ..., 5.900e-01, 1.560e+00,
         8.350e+02],
        [1.317e+01, 2.590e+00, 2.370e+00, ..., 6.000e-01, 1.620e+00,
         8.400e+02],
        [1.413e+01, 4.100e+00, 2.740e+00, ..., 6.100e-01, 1.600e+00,
         5.600e+02]]),
 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1

In [188]:
type(data)

sklearn.utils.Bunch

In [189]:
data_keys = data.keys()
data_keys

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names'])

In [190]:
pd.Series(data.target_names).value_counts()

class_0    1
class_2    1
class_1    1
dtype: int64

In [191]:
pd.Series(data.target).value_counts()

1    71
0    59
2    48
dtype: int64

In [192]:
data.DESCR.replace('\t\t', '').split('\n')

['.. _wine_dataset:',
 '',
 'Wine recognition dataset',
 '------------------------',
 '',
 '**Data Set Characteristics:**',
 '',
 '    :Number of Instances: 178 (50 in each of three classes)',
 '    :Number of Attributes: 13 numeric, predictive attributes and the class',
 '    :Attribute Information:',
 ' - Alcohol',
 ' - Malic acid',
 ' - Ash',
 '- Alcalinity of ash  ',
 ' - Magnesium',
 '- Total phenols',
 ' - Flavanoids',
 ' - Nonflavanoid phenols',
 ' - Proanthocyanins',
 '- Color intensity',
 ' - Hue',
 ' - OD280/OD315 of diluted wines',
 ' - Proline',
 '',
 '    - class:',
 '            - class_0',
 '            - class_1',
 '            - class_2',
 '',
 '    :Summary Statistics:',
 '    ',
 '    ============================= ==== ===== ======= =====',
 '                                   Min   Max   Mean     SD',
 '    ============================= ==== ===== ======= =====',
 '    Alcohol:                      11.0  14.8    13.0   0.8',
 '    Malic Acid:                   0.74 

Целевая переменная содержит 3 класса: class_0, class_1, class_2

In [193]:
X = pd.DataFrame(data=data.data, columns=data.feature_names)
X['target'] = data.target.astype('int64')
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 178 entries, 0 to 177
Data columns (total 14 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   alcohol                       178 non-null    float64
 1   malic_acid                    178 non-null    float64
 2   ash                           178 non-null    float64
 3   alcalinity_of_ash             178 non-null    float64
 4   magnesium                     178 non-null    float64
 5   total_phenols                 178 non-null    float64
 6   flavanoids                    178 non-null    float64
 7   nonflavanoid_phenols          178 non-null    float64
 8   proanthocyanins               178 non-null    float64
 9   color_intensity               178 non-null    float64
 10  hue                           178 non-null    float64
 11  od280/od315_of_diluted_wines  178 non-null    float64
 12  proline                       178 non-null    float64
 13  targe

Датафрейм имеет размерность 178 записей * 14 колонок, пропущенных значений не имеется

In [194]:
X

alcohol  malic_acid  ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71 2.43              15.60     127.00           2.80   
1      13.20        1.78 2.14              11.20     100.00           2.65   
2      13.16        2.36 2.67              18.60     101.00           2.80   
3      14.37        1.95 2.50              16.80     113.00           3.85   
4      13.24        2.59 2.87              21.00     118.00           2.80   
..       ...         ...  ...                ...        ...            ...   
173    13.71        5.65 2.45              20.50      95.00           1.68   
174    13.40        3.91 2.48              23.00     102.00           1.80   
175    13.27        4.28 2.26              20.00     120.00           1.59   
176    13.17        2.59 2.37              20.00     120.00           1.65   
177    14.13        4.10 2.74              24.50      96.00           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity  hue  \
0          3.06                  0.28             2.29             5.64 1.04   
1          2.76                  0.26             1.28             4.38 1.05   
2          3.24                  0.30             2.81             5.68 1.03   
3          3.49                  0.24             2.18             7.80 0.86   
4          2.69                  0.39             1.82             4.32 1.04   
..          ...                   ...              ...              ...  ...   
173        0.61                  0.52             1.06             7.70 0.64   
174        0.75                  0.43             1.41             7.30 0.70   
175        0.69                  0.43             1.35            10.20 0.59   
176        0.68                  0.53             1.46             9.30 0.60   
177        0.76                  0.56             1.35             9.20 0.61   

     od280/od315_of_diluted_wines  proline  target  
0                            3.92 1,065.00       0  
1                            3.40 1,050.00       0  
2                            3.17 1,185.00       0  
3                            3.45 1,480.00       0  
4                            2.93   735.00       0  
..                            ...      ...     ...  
173                          1.74   740.00       2  
174                          1.56   750.00       2  
175                          1.56   835.00       2  
176                          1.62   840.00       2  
177                          1.60   560.00       2  

[178 rows x 14 columns]

In [195]:
X_corr = X.corr()

In [196]:
high_corr=X_corr.query('0.5<abs(target)<1').index.to_list()
X_corr

alcohol  malic_acid   ash  alcalinity_of_ash  \
alcohol                          1.00        0.09  0.21              -0.31   
malic_acid                       0.09        1.00  0.16               0.29   
ash                              0.21        0.16  1.00               0.44   
alcalinity_of_ash               -0.31        0.29  0.44               1.00   
magnesium                        0.27       -0.05  0.29              -0.08   
total_phenols                    0.29       -0.34  0.13              -0.32   
flavanoids                       0.24       -0.41  0.12              -0.35   
nonflavanoid_phenols            -0.16        0.29  0.19               0.36   
proanthocyanins                  0.14       -0.22  0.01              -0.20   
color_intensity                  0.55        0.25  0.26               0.02   
hue                             -0.07       -0.56 -0.07              -0.27   
od280/od315_of_diluted_wines     0.07       -0.37  0.00              -0.28   
proline                          0.64       -0.19  0.22              -0.44   
target                          -0.33        0.44 -0.05               0.52   

                              magnesium  total_phenols  flavanoids  \
alcohol                            0.27           0.29        0.24   
malic_acid                        -0.05          -0.34       -0.41   
ash                                0.29           0.13        0.12   
alcalinity_of_ash                 -0.08          -0.32       -0.35   
magnesium                          1.00           0.21        0.20   
total_phenols                      0.21           1.00        0.86   
flavanoids                         0.20           0.86        1.00   
nonflavanoid_phenols              -0.26          -0.45       -0.54   
proanthocyanins                    0.24           0.61        0.65   
color_intensity                    0.20          -0.06       -0.17   
hue                                0.06           0.43        0.54   
od280/od315_of_diluted_wines       0.07           0.70        0.79   
proline                            0.39           0.50        0.49   
target                            -0.21          -0.72       -0.85   

                              nonflavanoid_phenols  proanthocyanins  \
alcohol                                      -0.16             0.14   
malic_acid                                    0.29            -0.22   
ash                                           0.19             0.01   
alcalinity_of_ash                             0.36            -0.20   
magnesium                                    -0.26             0.24   
total_phenols                                -0.45             0.61   
flavanoids                                   -0.54             0.65   
nonflavanoid_phenols                          1.00            -0.37   
proanthocyanins                              -0.37             1.00   
color_intensity                               0.14            -0.03   
hue                                          -0.26             0.30   
od280/od315_of_diluted_wines                 -0.50             0.52   
proline                                      -0.31             0.33   
target                                        0.49            -0.50   

                              color_intensity   hue  \
alcohol                                  0.55 -0.07   
malic_acid                               0.25 -0.56   
ash                                      0.26 -0.07   
alcalinity_of_ash                        0.02 -0.27   
magnesium                                0.20  0.06   
total_phenols                           -0.06  0.43   
flavanoids                              -0.17  0.54   
nonflavanoid_phenols                     0.14 -0.26   
proanthocyanins                         -0.03  0.30   
color_intensity                          1.00 -0.52   
hue                                     -0.52  1.00   
od280/od315_of_diluted_wines            -0.43  0.57   
proline                    

In [197]:
X=X.drop('target', axis=1)
X

alcohol  malic_acid  ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71 2.43              15.60     127.00           2.80   
1      13.20        1.78 2.14              11.20     100.00           2.65   
2      13.16        2.36 2.67              18.60     101.00           2.80   
3      14.37        1.95 2.50              16.80     113.00           3.85   
4      13.24        2.59 2.87              21.00     118.00           2.80   
..       ...         ...  ...                ...        ...            ...   
173    13.71        5.65 2.45              20.50      95.00           1.68   
174    13.40        3.91 2.48              23.00     102.00           1.80   
175    13.27        4.28 2.26              20.00     120.00           1.59   
176    13.17        2.59 2.37              20.00     120.00           1.65   
177    14.13        4.10 2.74              24.50      96.00           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity  hue  \
0          3.06                  0.28             2.29             5.64 1.04   
1          2.76                  0.26             1.28             4.38 1.05   
2          3.24                  0.30             2.81             5.68 1.03   
3          3.49                  0.24             2.18             7.80 0.86   
4          2.69                  0.39             1.82             4.32 1.04   
..          ...                   ...              ...              ...  ...   
173        0.61                  0.52             1.06             7.70 0.64   
174        0.75                  0.43             1.41             7.30 0.70   
175        0.69                  0.43             1.35            10.20 0.59   
176        0.68                  0.53             1.46             9.30 0.60   
177        0.76                  0.56             1.35             9.20 0.61   

     od280/od315_of_diluted_wines  proline  
0                            3.92 1,065.00  
1                            3.40 1,050.00  
2                            3.17 1,185.00  
3                            3.45 1,480.00  
4                            2.93   735.00  
..                            ...      ...  
173                          1.74   740.00  
174                          1.56   750.00  
175                          1.56   835.00  
176                          1.62   840.00  
177                          1.60   560.00  

[178 rows x 13 columns]

In [198]:
temp = X[high_corr].apply(lambda x: x**2)
temp.columns = [el+'_2' for el in list(X[high_corr])]
X=X.join(temp)

In [199]:
X

alcohol  malic_acid  ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71 2.43              15.60     127.00           2.80   
1      13.20        1.78 2.14              11.20     100.00           2.65   
2      13.16        2.36 2.67              18.60     101.00           2.80   
3      14.37        1.95 2.50              16.80     113.00           3.85   
4      13.24        2.59 2.87              21.00     118.00           2.80   
..       ...         ...  ...                ...        ...            ...   
173    13.71        5.65 2.45              20.50      95.00           1.68   
174    13.40        3.91 2.48              23.00     102.00           1.80   
175    13.27        4.28 2.26              20.00     120.00           1.59   
176    13.17        2.59 2.37              20.00     120.00           1.65   
177    14.13        4.10 2.74              24.50      96.00           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity  hue  \
0          3.06                  0.28             2.29             5.64 1.04   
1          2.76                  0.26             1.28             4.38 1.05   
2          3.24                  0.30             2.81             5.68 1.03   
3          3.49                  0.24             2.18             7.80 0.86   
4          2.69                  0.39             1.82             4.32 1.04   
..          ...                   ...              ...              ...  ...   
173        0.61                  0.52             1.06             7.70 0.64   
174        0.75                  0.43             1.41             7.30 0.70   
175        0.69                  0.43             1.35            10.20 0.59   
176        0.68                  0.53             1.46             9.30 0.60   
177        0.76                  0.56             1.35             9.20 0.61   

     od280/od315_of_diluted_wines  proline  alcalinity_of_ash_2  \
0                            3.92 1,065.00               243.36   
1                            3.40 1,050.00               125.44   
2                            3.17 1,185.00               345.96   
3                            3.45 1,480.00               282.24   
4                            2.93   735.00               441.00   
..                            ...      ...                  ...   
173                          1.74   740.00               420.25   
174                          1.56   750.00               529.00   
175                          1.56   835.00               400.00   
176                          1.62   840.00               400.00   
177                          1.60   560.00               600.25   

     total_phenols_2  flavanoids_2  hue_2  od280/od315_of_diluted_wines_2  \
0               7.84          9.36   1.08                           15.37   
1               7.02          7.62   1.10                           11.56   
2               7.84         10.50   1.06                           10.05   
3              14.82         12.18   0.74                           11.90   
4               7.84          7.24   1.08                            8.58   
..               ...           ...    ...                             ...   
173             2.82          0.37   0.41                            3.03   
174             3.24          0.56   0.49                            2.43   
175             2.53          0.48   0.35                            2.43   
176             2.72          0.46   0.36                            2.62   
177             4.20          0.58   0.37                            2.56   

       proline_2  
0   1,134,225.00  
1   1,102,500.00  
2   1,404,225.00  
3   2,190,400.00  
4     540,225.00  
..           ...  
173   547,600.00  
174   562,500.00  
175   697,225.00  
176   705,600.00  
177   313,600.00  

[178 rows x 19 columns]

In [200]:
X.describe()

alcohol  malic_acid    ash  alcalinity_of_ash  magnesium  \
count   178.00      178.00 178.00             178.00     178.00   
mean     13.00        2.34   2.37              19.49      99.74   
std       0.81        1.12   0.27               3.34      14.28   
min      11.03        0.74   1.36              10.60      70.00   
25%      12.36        1.60   2.21              17.20      88.00   
50%      13.05        1.87   2.36              19.50      98.00   
75%      13.68        3.08   2.56              21.50     107.00   
max      14.83        5.80   3.23              30.00     162.00   

       total_phenols  flavanoids  nonflavanoid_phenols  proanthocyanins  \
count         178.00      178.00                178.00           178.00   
mean            2.30        2.03                  0.36             1.59   
std             0.63        1.00                  0.12             0.57   
min             0.98        0.34                  0.13             0.41   
25%             1.74        1.21                  0.27             1.25   
50%             2.35        2.13                  0.34             1.56   
75%             2.80        2.88                  0.44             1.95   
max             3.88        5.08                  0.66             3.58   

       color_intensity    hue  od280/od315_of_diluted_wines  proline  \
count           178.00 178.00                        178.00   178.00   
mean              5.06   0.96                          2.61   746.89   
std               2.32   0.23                          0.71   314.91   
min               1.28   0.48                          1.27   278.00   
25%               3.22   0.78                          1.94   500.50   
50%               4.69   0.96                          2.78   673.50   
75%               6.20   1.12                          3.17   985.00   
max              13.00   1.71                          4.00 1,680.00   

       alcalinity_of_ash_2  total_phenols_2  flavanoids_2  hue_2  \
count               178.00           178.00        178.00 178.00   
mean                391.14             5.66          5.11   0.97   
std                 133.67             2.94          4.21   0.44   
min                 112.36             0.96          0.12   0.23   
25%                 295.84             3.04          1.45   0.61   
50%                 380.25             5.55          4.56   0.93   
75%                 462.25             7.84          8.27   1.25   
max                 900.00            15.05         25.81   2.92   

       od280/od315_of_diluted_wines_2    proline_2  
count                          178.00       178.00  
mean                             7.32   656,459.14  
std                              3.58   555,859.14  
min                              1.61    77,284.00  
25%                              3.75   250,501.00  
50%                              7.73   453,604.50  
75%                             10.05   970,225.00  
max                             16.00 2,822,400.00